pkg/hubble/relay/observer.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package relay
+
+import (
+	"context"
+	"io"
+	"time"
+
+	observerpb "github.com/cilium/cilium/api/v1/observer"
+	"github.com/sirupsen/logrus"
+
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// ensure that Server implements the observer.ObserverServer interface.
+var _ observerpb.ObserverServer = (*Server)(nil)
+
+func newObserverClient(target string, dialTimeout time.Duration) (observerpb.ObserverClient, *grpc.ClientConn, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), dialTimeout)
+	defer cancel()
+	//FIXME: remove WithInsecure once mutual TLS is implemented
+	conn, err := grpc.DialContext(ctx, target, grpc.WithInsecure(), grpc.WithBlock())
+	if err != nil {
+		return nil, nil, err
+	}
+	return observerpb.NewObserverClient(conn), conn, nil
+}
+
+// GetFlows implements observer.ObserverServer.GetFlows by proxying requests to
+// the hubble instance the proxy is connected to.
+func (s *Server) GetFlows(req *observerpb.GetFlowsRequest, stream observerpb.Observer_GetFlowsServer) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	g, ctx := errgroup.WithContext(ctx)
+	g.Go(func() error {
+		select {
+		case <-s.stop:
+			cancel()
+		case <-ctx.Done():
+		}
+		return nil
+	})
+
+	//TODO: figure out what makes a reasonnable channel size
+	peers := s.peerList()
+	flows := make(chan *observerpb.GetFlowsResponse, 10*len(peers))
+	for _, p := range peers {
+		p := p
+		g.Go(func() error {
+			//FIXME: establishing a new connection to every peer for every
+			// request is costly. Possible solutions:
+			// - selective connection: only connect to peers that have the
+			//   requested information (it still has the cost of connect/disconnect per
+			//   GetFlows request but it's at least limited in terms of scope).
+			// - Maintain a connection with all peers and reconnect when necessary.
```

Comment:
(no action required)

I think this is the option we want to pursue in the long run. gRPC client connections seem thread-safe, so reusing the connections to the selected peers seems to be the way to go.